clang/include/clang/Sema/Sema.h

```diff
@@ -12055,6 +12055,11 @@ class Sema final {
   // useful notes that shows where the kernel was called.
   bool ConstructingOpenCLKernel = false;
 
+  /// FunctionDecls and SourceLocations for which checkSYCLDeviceFunction has
+  /// emitted a (maybe deferred) "bad call" diagnostic.  We use this to avoid
+  /// emitting the same deferred diag twice.
```

Comment:
Do you know why the diagnostic is emitted twice?

Reply:
First, functiondecl and location are insufficient for this.  For example, 'notes' can happen multiple times on the same location, You'd also need diagnostic id, though even that I believe is insufficient.

Second, please don't try to solve this problem yet.  I've been working with a couple of people in LLVM community to attempt to fix this problem in a more correct way that actually solves the problem throughout the compiler. In the meantime, i think we have to put up with doubled-diagnostics until we can get it fixed.

Reply:
@erichkeane 
Do you suggest use undeferred diagnostic?

Reply:
I think we should use the deferred diagnostics, because it is our intent to switch everything to the deferred system over time.

Reply:
I suppose Erich suggests do not try to fix double diagnostics, because it's known clang bug...

Reply:
> 
> 
> I suppose Erich suggests do not try to fix double diagnostics, because it's known clang bug...

Yep, thats exactly it.  Do the deferred diagnostic and 'live with' the duplicates until Richard, Aaron and I solve the problem (or at least create the infrastructure to do so).

Reply:
Why this is still here?

Reply:
Removed